# PART I: DATA WRANGLING (FCB LALIGA MATCHES 2004-2021)

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#collection">Data Collection</a></li>
<li><a href="#assess">Assesment</a></li>
<li><a href="#clean">Data Cleaning</a></li>
<li><a href="#cleaned_data">Cleaned FCB Dataset</a></li>
</ul>

In [1]:
#imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import glob
import os

In [2]:
#change the settings for good output
pd.set_option('display.max_rows',100)
pd.set_option('display.max_columns',20)
pd.set_option('display.max_colwidth',200)

<a id='collection'></a>
## Data Collection

In [3]:
#read the file names and join with its extension (.json)
path = 'C:\Data-Analysis-projects\FC Barcelona Data Analysis\Datasets'
file_name = glob.glob(os.path.join(path,'*.json'))
print(file_name)

['C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20042005-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20052006-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20062007-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20072008-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20082009-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20092010-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20102011-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20112012-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20122013-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20132014-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data An

In [4]:
#read the file in pandas and concatenate them to a single dataframe
dfs = []
for file in file_name:
    df = pd.read_json(file)
    dfs.append(df)
    
fcb = pd.concat(dfs, axis=0, ignore_index=True)

In [5]:
#first 5 data
fcb.head(3)

match_id  match_date      kick_off  \
0     68314  2004-12-04  20:00:00.000   
1     68313  2004-10-24  21:00:00.000   
2     68316  2005-05-01  19:00:00.000   

                                                                      competition  \
0  {'competition_id': 11, 'country_name': 'Spain', 'competition_name': 'La Liga'}   
1  {'competition_id': 11, 'country_name': 'Spain', 'competition_name': 'La Liga'}   
2  {'competition_id': 11, 'country_name': 'Spain', 'competition_name': 'La Liga'}   

                                          season  \
0  {'season_id': 37, 'season_name': '2004/2005'}   
1  {'season_id': 37, 'season_name': '2004/2005'}   
2  {'season_id': 37, 'season_name': '2004/2005'}   

                                                                                                                                                                                                 home_team  \
0  {'home_team_id': 217, 'home_team_name': 'Barcelona', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4953, 'name': 'Frank Rijkaard...   
1  {'home_team_id': 217, 'home_team_name': 'Barcelona', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4953, 'name': 'Frank Rijkaard...   
2  {'home_team_id': 217, 'home_team_name': 'Barcelona', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4953, 'name': 'Frank Rijkaard...   

                                                                                                                                                                                                 away_team  \
0  {'away_team_id': 223, 'away_team_name': 'Málaga', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4994, 'name': 'Gregorio Manzano ...   
1  {'away_team_id': 422, 'away_team_name': 'Osasuna', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 3196, 'name': 'Javier Aguirre O...   
2  {'away_team_id': 608, 'away_team_name': 'Albacete', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 6233, 'name': 'Martín Monteagu...   

   home_score  away_score match_status match_status_360      last_updated  \
0           4           0    available        scheduled  2020-07-29T05:00   
1           3           0    available        scheduled  2020-07-29T05:00   
2           2           0    available        scheduled  2020-07-29T05:00   

          last_updated_360  \
0  2021-06-13T16:17:31.694   
1  2021-06-13T16:17:31.694   
2  2021-06-13T16:17:31.694   

                                                                              metadata  \
0  {'data_version': '1.1.0', 'shot_fidelity_version': '2', 'xy_fidelity_version': '2'}   
1  {'data_version': '1.1.0', 'shot_fidelity_version': '2', 'xy_fidelity_version': '2'}   
2  {'data_version': '1.1.0', 'shot_fidelity_version': '2', 'xy_fidelity_version': '2'}   

   match_week                    competition_stage  \
0          14  {'id': 1, 'name': 'Regular Season'}   
1           8  {'id': 1, 'name': 'Regular Season'}   
2          34  {'id': 1, 'name': 'Regular Season'}   

                                                                    stadium  \
0  {'id': 342, 'name': 'Camp Nou', 'country': {'id': 214, 'name': 'Spain'}}   
1  {'id': 342, 'name': 'Camp Nou', 'country': {'id': 214, 'name': 'Spain'}}   
2  {'id': 342, 'name': 'Camp Nou', 'country': {'id': 214, 'name': 'Spain'}}   

                                                                                       referee  
0           {'id': 993, 'name': 'José Omar Losantos', 'country': {'id': 214, 'name': 'Spain'}}  
1  {'id': 994, 'name': 'Vicente José Lizondo Cortés', 'country': {'id': 214, 'name': 'Spain'}}  
2     {'id': 2535, 'na

In [6]:
#basic info like the shape, datatypes and null values
fcb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 520 entries, 0 to 519
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   match_id           520 non-null    int64 
 1   match_date         520 non-null    object
 2   kick_off           520 non-null    object
 3   competition        520 non-null    object
 4   season             520 non-null    object
 5   home_team          520 non-null    object
 6   away_team          520 non-null    object
 7   home_score         520 non-null    int64 
 8   away_score         520 non-null    int64 
 9   match_status       520 non-null    object
 10  match_status_360   520 non-null    object
 11  last_updated       520 non-null    object
 12  last_updated_360   520 non-null    object
 13  metadata           520 non-null    object
 14  match_week         520 non-null    int64 
 15  competition_stage  520 non-null    object
 16  stadium            520 non-null    object
 1

> The dataset has 520 matches played with 18 columns information, only referee has missing values and some datatypes need to change

In [7]:
# descriptive statistics
fcb.describe()

match_id  home_score  away_score  match_week
count  5.200000e+02  520.000000  520.000000  520.000000
mean   3.744410e+05    1.982692    1.500000   19.457692
std    9.194129e+05    1.723041    1.492434   10.777579
min    9.575000e+03    0.000000    0.000000    1.000000
25%    6.915675e+04    1.000000    0.000000   10.000000
50%    6.932250e+04    2.000000    1.000000   20.000000
75%    2.667258e+05    3.000000    2.000000   29.000000
max    3.773695e+06    8.000000    8.000000   38.000000

> We can see the maximum score is 8 goals, minimum is obviously zero, and an average of 2 goals precisely. 

In [8]:
#check for duplicates
fcb['match_id'].duplicated().any()

False

In [9]:
#check the competition column
fcb['competition'].value_counts()

{'competition_id': 11, 'country_name': 'Spain', 'competition_name': 'La Liga'}    520
Name: competition, dtype: int64

> it's has only a single value object. it should be discarded

In [10]:
#check the season column
fcb['season'].value_counts()

{'season_id': 26, 'season_name': '2014/2015'}    38
{'season_id': 23, 'season_name': '2011/2012'}    37
{'season_id': 1, 'season_name': '2017/2018'}     36
{'season_id': 90, 'season_name': '2020/2021'}    35
{'season_id': 21, 'season_name': '2009/2010'}    35
{'season_id': 4, 'season_name': '2018/2019'}     34
{'season_id': 2, 'season_name': '2016/2017'}     34
{'season_id': 42, 'season_name': '2019/2020'}    33
{'season_id': 22, 'season_name': '2010/2011'}    33
{'season_id': 27, 'season_name': '2015/2016'}    33
{'season_id': 24, 'season_name': '2012/2013'}    32
{'season_id': 25, 'season_name': '2013/2014'}    31
{'season_id': 41, 'season_name': '2008/2009'}    31
{'season_id': 40, 'season_name': '2007/2008'}    28
{'season_id': 39, 'season_name': '2006/2007'}    26
{'season_id': 38, 'season_name': '2005/2006'}    17
{'season_id': 37, 'season_name': '2004/2005'}     7
Name: season, dtype: int64

> we will retain only the season_name and discard the remaining data

In [11]:
#inspect home team and team
fcb[['home_team','away_team']].sample(3)

home_team  \
83   {'home_team_id': 1217, 'home_team_name': 'Racing Santander', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 2, 'name': 'Juan Ramó...   
18   {'home_team_id': 1043, 'home_team_name': 'Mallorca', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 486, 'name': 'Héctor Raúl Cúp...   
151  {'home_team_id': 217, 'home_team_name': 'Barcelona', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 36, 'name': 'Josep Guardiola ...   

                                                                                                                                                                                                   away_team  
83   {'away_team_id': 217, 'away_team_name': 'Barcelona', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 36, 'name': 'Josep Guardiola ...  
18   {'away_team_id': 217, 'away_team_name': 'Barcelona', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4953, 'name': 'Frank Rijkaard...  
151  {'away_team_id': 1217, 'away_team_name': 'Racing Santander', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4954, 'name': 'Miguel...

> we will extract home_team_name and home_manager likewise away_team_name and away_team-manager

In [12]:
#inspect match_status
fcb['match_status'].value_counts() 

available    520
Name: match_status, dtype: int64

> match status is unique, we will discard it

In [13]:
#inspect match_status_360
fcb['match_status_360'].value_counts()

scheduled     485
available      34
processing      1
Name: match_status_360, dtype: int64

In [14]:
#inspect meta data
fcb['metadata'].value_counts()

{'data_version': '1.1.0', 'shot_fidelity_version': '2', 'xy_fidelity_version': '2'}    465
{'data_version': '1.1.0', 'shot_fidelity_version': '2'}                                 26
{'data_version': '1.0.2', 'shot_fidelity_version': '2'}                                 16
{'data_version': '1.0.3', 'shot_fidelity_version': '2'}                                 13
Name: metadata, dtype: int64

> Not important for our analysis

In [15]:
#inspect competition stage
fcb['competition_stage'].value_counts()

{'id': 1, 'name': 'Regular Season'}    520
Name: competition_stage, dtype: int64

> has unique values, it shall be discarded

In [16]:
#inspect stadium
fcb['stadium'].value_counts().head()

{'id': 342, 'name': 'Camp Nou', 'country': {'id': 214, 'name': 'Spain'}}                         262
{'id': 361, 'name': 'Estadio de la Cerámica', 'country': {'id': 214, 'name': 'Spain'}}            15
{'id': 353, 'name': 'Estadio Santiago Bernabéu', 'country': {'id': 214, 'name': 'Spain'}}         15
{'id': 344, 'name': 'Estadio de Mestalla', 'country': {'id': 214, 'name': 'Spain'}}               15
{'id': 349, 'name': 'Estadio Ramón Sánchez Pizjuán', 'country': {'id': 214, 'name': 'Spain'}}     14
Name: stadium, dtype: int64

> we will extract only the stadium name

In [17]:
#inspect referee
fcb['referee'].value_counts().head()

{'id': 221, 'name': 'Alberto Undiano Mallenco', 'country': {'id': 214, 'name': 'Spain'}}      33
{'id': 180, 'name': 'Antonio Miguel Mateu Lahoz', 'country': {'id': 214, 'name': 'Spain'}}    32
{'id': 222, 'name': 'David Fernández Borbalan', 'country': {'id': 214, 'name': 'Spain'}}      30
{'id': 1008, 'name': 'Carlos Clos Gómez', 'country': {'id': 214, 'name': 'Spain'}}            20
{'id': 1003, 'name': 'Fernando Teixeira', 'country': {'id': 214, 'name': 'Spain'}}            20
Name: referee, dtype: int64

> we will extract only the referee name

<a id='assess'></a>
## Assessment
1. match_date, kick_off should be datetime
2. competition, competition_stafe, and match_status are unique (not important for our analysis) and last_updated, last_updated_360, and meta_data are not needed.
3. season should have only the season_name (For example, 2004/2005)
4. home_team and away_team should consist of only home_team name, also we can extract manager from it.
5. stadium and referee should contain only the name
6. create new column 'match_outcome' that have values of win, draw, lose for FCB.

In [18]:
#copy the data before cleaning
fcb_df = fcb.copy()

<a id='clean'></a>
## Data Cleaning

**Issue #1:** match_date and kick_off should be datetime
#### Define
Use `.astype()` to convert it to datetime datatype.
#### Code

In [19]:
fcb_df['match_date'] = fcb_df['match_date'].astype('datetime64')
fcb_df['kick_off'] = fcb_df['kick_off'].astype('datetime64').dt.time

#### Test

In [20]:
fcb_df[['match_date','kick_off']].dtypes

match_date    datetime64[ns]
kick_off              object
dtype: object

**Issue #2:** competition, competition_stage, and match_status are unique (not important for our analysis) and last_updated, last_updated_360, match_status_360 and metadata are not needed.
#### Define
Use `.drop()` to discard the unwanted columns.
#### Code

In [21]:
cols_drop = ['competition','competition_stage','match_status','last_updated','last_updated_360','metadata','match_status_360']
fcb_df.drop(cols_drop, axis=1, inplace=True)

#### Test

In [22]:
assert ~fcb_df.columns.isin(cols_drop).any()

**Issue #3:** season should have only the season_name (For example, 2004/2005)
#### Define
Use the key attribute pairs of json and extract the season name, and drop the competition column.
#### Code

In [23]:
fcb_df['season'] = fcb_df['season'].str.get('season_name')

#### Test

In [24]:
fcb_df['season'].value_counts()

2014/2015    38
2011/2012    37
2017/2018    36
2020/2021    35
2009/2010    35
2018/2019    34
2016/2017    34
2019/2020    33
2010/2011    33
2015/2016    33
2012/2013    32
2013/2014    31
2008/2009    31
2007/2008    28
2006/2007    26
2005/2006    17
2004/2005     7
Name: season, dtype: int64

**Issue #4:** home_team and away_team should consist of only home_team name, also we can extract managers from it.
#### Define
Use the key attribute pairs of json and extract the team and its managers. discard the unwanted values.
#### Code

In [25]:
def extract_team_managers(col,df=fcb_df):
    '''
    This function extract team and its manager from col.
    '''
    managers = []
    for i in range(df.shape[0]):
        if df[col].str.get('managers')[i] is None:
            managers.append(np.NaN)
        else:
            managers.append(df[col].str.get('managers')[i][0].get('name'))
    #assign the array to dataframe
    df[col+'_managers'] = managers
    
    #home team
    df[col] = df[col].str.get(col+'_name')
    

In [26]:
extract_team_managers('home_team')
extract_team_managers('away_team')

#### Test

In [27]:
fcb_df[['home_team','home_team_managers','away_team','away_team_managers']].head()

home_team          home_team_managers  away_team  \
0  Barcelona              Frank Rijkaard     Málaga   
1  Barcelona              Frank Rijkaard    Osasuna   
2  Barcelona              Frank Rijkaard   Albacete   
3  Barcelona              Frank Rijkaard    Levante   
4   Albacete  José Manuel González López  Barcelona   

             away_team_managers  
0  Gregorio Manzano Ballesteros  
1       Javier Aguirre Onaindía  
2  Martín Monteagudo Monteagudo  
3                Bernd Schuster  
4                Frank Rijkaard

**Issue #5:** stadium and referee should contain only the name
#### Define
Use the key attribute pairs of json and extract the stadium nameand referee name. discrad the unwanted values.
#### Code

In [28]:
fcb_df['stadium'] = fcb_df['stadium'].str.get('name')
fcb_df['referee'] = fcb_df['referee'].str.get('name')

#### Test

In [29]:
fcb_df[['stadium','referee']].head()

stadium                      referee
0                 Camp Nou           José Omar Losantos
1                 Camp Nou  Vicente José Lizondo Cortés
2                 Camp Nou      Carlos Velasco Carballo
3                 Camp Nou     David Fernández Borbalan
4  Estadio Carlos Belmonte        Alfonso Perez Burrull

**Issue #6:** create new column 'match_outcome' that have values of win, draw, lose for FCB.
#### Define
Use filtering and conditional statements to create new match_outcome
#### Code

In [30]:
mask1 = fcb_df.query(" (home_team == 'Barcelona' and home_score > away_score) or (away_team == 'Barcelona' and home_score < away_score)").index
mask2 = fcb_df.query("(home_team != 'Barcelona' and home_score > away_score) or (away_team != 'Barcelona' and home_score < away_score)").index
mask3 = fcb_df.query('home_score == away_score').index
fcb_df.loc[mask1,'match_outcome'] = 'Win'
fcb_df.loc[mask2,'match_outcome'] = 'Lost'
fcb_df.loc[mask3,'match_outcome'] = 'Draw'


### Test

In [31]:
fcb_df.match_outcome.sample(10)

367     Win
129    Draw
251    Lost
60     Lost
346     Win
155    Lost
93     Lost
307     Win
453     Win
219     Win
Name: match_outcome, dtype: object

In [32]:
fcb_df.match_outcome.value_counts()

Win     383
Draw     86
Lost     51
Name: match_outcome, dtype: int64

<a id='cleaned_data'></a>
## Cleaned FCB Dataset

In [33]:
fcb_df.head()

match_id match_date  kick_off     season  home_team  away_team  home_score  \
0     68314 2004-12-04  20:00:00  2004/2005  Barcelona     Málaga           4   
1     68313 2004-10-24  21:00:00  2004/2005  Barcelona    Osasuna           3   
2     68316 2005-05-01  19:00:00  2004/2005  Barcelona   Albacete           2   
3     68315 2004-12-21  20:00:00  2004/2005  Barcelona    Levante           2   
4     69153 2004-12-11  20:00:00  2004/2005   Albacete  Barcelona           1   

   away_score  match_week                  stadium  \
0           0          14                 Camp Nou   
1           0           8                 Camp Nou   
2           0          34                 Camp Nou   
3           1          17                 Camp Nou   
4           2          15  Estadio Carlos Belmonte   

                       referee          home_team_managers  \
0           José Omar Losantos              Frank Rijkaard   
1  Vicente José Lizondo Cortés              Frank Rijkaard   
2      Carlos Velasco Carballo              Frank Rijkaard   
3     David Fernández Borbalan              Frank Rijkaard   
4        Alfonso Perez Burrull  José Manuel González López   

             away_team_managers match_outcome  
0  Gregorio Manzano Ballesteros           Win  
1       Javier Aguirre Onaindía           Win  
2  Martín Monteagudo Monteagudo           Win  
3                Bernd Schuster           Win  
4                Frank Rijkaard           Win

In [34]:
#fcb_df.to_csv('./Datasets/fcb2004-2021.csv', index=False)

# Part II: Exploratory Data Analysis

In [35]:
fcb_df.match_outcome.value_counts()

Win     383
Draw     86
Lost     51
Name: match_outcome, dtype: int64